src/groupby/hash/groupby_compute_api.h

```diff
@@ -0,0 +1,210 @@
+  //extract_groupby_result<<<extract_grid_size, block_size>>>(the_map.get(),
+  //                                                          the_map->size(),
+  //                                                          out_groupby_column,
+  //                                                          out_aggregation_column,
+  //                                                          global_write_index);
+  
+  // FIXME Work around for above kernel failing to launch for some instantiations of the_map template class
+  map_type * map = the_map.get();
+  typename map_type::size_type map_size = the_map->size();
+  void *args[] = { &map, &map_size, &out_groupby_column, &out_aggregation_column, &global_write_index};
+
+  void (*func)(const map_type * const, 
+               const typename map_type::size_type, 
+               typename map_type::key_type * const, 
+               typename map_type::mapped_type * const,
+               unsigned int * const ) = &(extract_groupby_result<map_type>);
+
+  CUDA_RT_CALL(cudaLaunchKernel((const void*) func, extract_grid_size, block_size, args, 0, 0));
+  // FIXME End work around
+
+  CUDA_RT_CALL(cudaDeviceSynchronize());
+
+  // At the end of the extraction kernel, the global write index will be equal to
+  // the size of the output. Update the output size.
+  *out_size = *global_write_index;
+  CUDA_RT_CALL(cudaFree(global_write_index));
+
+  // Optionally sort the groupby/aggregation result columns
+  if(true == sort_result)
+  {
+    // Allocate double buffers needed for the cub Radix Sort
+    groupby_type * groupby_result_alt;
+    CUDA_RT_CALL(cudaMalloc(&groupby_result_alt, *out_size * sizeof(groupby_type)));
+
+    aggregation_type * aggregation_result_alt;
+    CUDA_RT_CALL(cudaMalloc(&aggregation_result_alt, *out_size * sizeof(aggregation_type)));
+
+    cub::DoubleBuffer<groupby_type>     d_keys(out_groupby_column, groupby_result_alt);
+    cub::DoubleBuffer<aggregation_type> d_vals(out_aggregation_column, aggregation_result_alt);
+
+    // When called with temp_storage == nullptr, simply returns the required allocation size in
+    // temp_storage_bytes
+    void *d_temp_storage = nullptr;
+    size_t temp_storage_bytes = 0;
+    CUDA_RT_CALL(cub::DeviceRadixSort::SortPairs(d_temp_storage, temp_storage_bytes, d_keys, d_vals, *out_size));
+    
+    // allocate temp storage here and call sort again to actually sort arrays
+    CUDA_RT_CALL(cudaMalloc(&d_temp_storage, temp_storage_bytes));
+    CUDA_RT_CALL(cub::DeviceRadixSort::SortPairs(d_temp_storage, temp_storage_bytes, d_keys, d_vals, *out_size));
+
+    CUDA_RT_CALL(cudaDeviceSynchronize());
+
+    // Update output pointers with sorted result
+    // TODO Find a better way to do this. 
+    // Sorted output may be in a different buffer than what was originally passed in... need to copy it 
+    CUDA_RT_CALL(cudaMemcpy(out_groupby_column, d_keys.Current(), *out_size * sizeof(groupby_type), cudaMemcpyDefault));
```

Comment:
Since we aren't using streams these cudaMemcpy are synchronous calls and we are synchronizing the whole device just a few lines before.  This would really limit our ability to leverage libgdf properly.

Reply:
See above comment. This sort call of the result will go away when I implement multi-column groupbys.